forum/qa-plugin/q2a-googleauthenticator-login/q2a-googleauthenticator-overrides.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+function qa_set_logged_in_user($userId, $handle = '', $remember = false, $source = null)
+{
+    require_once QA_INCLUDE_DIR . 'app/cookies.php';
+    qa_start_session();
+    if (!isset($userId)) {
+        // logout
+        qa_report_event(
+            'u_logout',
+            qa_get_logged_in_userid(),
+            qa_get_logged_in_handle(),
+            qa_cookie_get()
+        );
+        qa_clear_session_cookie();
+        qa_clear_session_user();
+
+        return;
+    }
+
+    // login
+    require_once QA_INCLUDE_DIR . 'db/selects.php';
+    $result = qa_db_read_all_assoc(
+        qa_db_query_sub(
+            'SELECT 2fa_enabled FROM ^users WHERE userid = #', $userId
+        )
+    );
+    if (count($result) !== 1) {
+        echo 'Invalid num_rows';
+        die;
+    }
+    $usingTwoFactorAuth = (bool) $result[0]['2fa_enabled'];
+    if ($usingTwoFactorAuth && '2fa' !== $source) {
+        echo '
+<form method="post" id="form" action="./2fa-auth">
```

Comment:
Atrybut `id` o nazwie "form" jest zbyt ogólny - mógłby dotyczyć każdego formularza na stronie. Lepiej nadać mu nazwę od roli, którą pełni, np. "authForm".

Reply:
Spoko. Ten PR został mi podkradziony :P ja miałem dość mało tam zrobione i generalnie dużo się zmieniło od mojej pierwotnej wizji tej funkcjonalności. Przyjąłem jakoś stan zastany i doprowadziłem do używalności. Na chwilę obecną było istotne uruchomienie tego, doprowadzenie do działania bez błędów. 

Muszę podkreślić, że nie jest to kod który widziałem w swojej wizji, ale najistotniejsze jest to, że core został wdrożony, czyli ktoś odwalił za mnie robotę :)  zaraz poprawię w wolnej chwili :)

Reply:
Ok, nie wnikałem kto komu co podkradł. ;)
Dziwi mnie trochę, dlaczego bezpośrednio pod formularzem jest skrypt, który go submituje - czy wysłanie nie powinno być na akcję użytkownika?

Reply:
Szczerze - nie mam pojęcia. W tej chwili jka pobierzesz ten kod, to przy włączonym pluginie i uruchomionej 2fa na użytkowniku możesz spróbować się zalogować. Objawi się to tym, że przy wrzuceniu kredek do logowania przekieruje Cię najpierw na białą stronę z napisem "Trwa przekierowywanie..." a następnie do formularza z inputem na cyferki. Z mojego punktu widzenia to jest bardzo dziwne zachowanie i pierwsze wrażenie jakie miałem to "wygląda jak hijacking". 

Być może uda się inaczej. Ale to wymaga wolnego weekendu :)